presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    @VisibleForTesting
+    public synchronized void collectDynamicFilters()
+    {
+        List<QueryId> queriesWithDynamicFilters = dynamicFilterSummaries.entrySet()
+                .stream()
+                .filter(entry -> !entry.getValue().isPresent())
+                .map(entry -> entry.getKey().getQueryId())
+                .distinct()
+                .collect(toImmutableList());
+
+        for (QueryId queryId : queriesWithDynamicFilters) {
+            Supplier<List<StageInfo>> stageInfoSupplier = queries.get(queryId);
+            if (stageInfoSupplier != null) {
+                for (StageInfo stageInfo : stageInfoSupplier.get()) {
+                    StageState stageState = stageInfo.getState();
+                    // wait until stage has finished scheduling tasks
+                    if (!stageState.noMoreTasks()) {
+                        continue;
+                    }
+                    List<TaskInfo> tasks = stageInfo.getTasks();
+                    Map<String, List<Domain>> stageDynamicFilterDomains = tasks.stream()
+                            .map(taskInfo -> taskInfo.getTaskStatus().getDynamicFilterDomains())
+                            .flatMap(taskDynamicTupleDomains -> taskDynamicTupleDomains.entrySet().stream())
+                            .collect(groupingBy(Map.Entry::getKey, mapping(Map.Entry::getValue, toList())));
+
+                    stageDynamicFilterDomains.entrySet().stream()
+                            // check if all tasks of a dynamic filter source have reported dynamic filter summary
+                            .filter(entry -> entry.getValue().size() == tasks.size())
+                            .forEach(entry -> putDomainSummary(queryId, entry.getKey(), Domain.union(entry.getValue())));
+                }
+            }
+        }
+    }
+
+    private void putDomainSummary(QueryId queryId, String filterId, Domain summary)
+    {
+        dynamicFilterSummaries.put(SourceDescriptor.of(queryId, filterId), Optional.of(summary));
+    }
+
+    @VisibleForTesting
+    public Optional<Domain> getSummary(QueryId queryId, String filterId)
+    {
+        return dynamicFilterSummaries.computeIfAbsent(SourceDescriptor.of(queryId, filterId), ignored -> Optional.empty());
+    }
+
+    public synchronized void removeQuery(QueryId queryId)
+    {
+        dynamicFilterSummaries.keySet().removeIf(sourceDescriptor -> sourceDescriptor.getQueryId().equals(queryId));
+        queries.remove(queryId);
+    }
+
+    public Supplier<TupleDomain<ColumnHandle>> createDynamicFilterSupplier(QueryId queryId, List<DynamicFilters.Descriptor> dynamicFilters, Map<Symbol, ColumnHandle> columnHandles)
+    {
+        Map<String, ColumnHandle> sourceColumnHandles = extractSourceExpressionSymbols(dynamicFilters)
+                .entrySet().stream()
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> columnHandles.get(entry.getValue())));
+
+        return () -> {
```

Comment:
I think this can be more streamlined:
```
dynamicFilters.stream()
  .map(filter -> getSummary(queryId, dynamicFilter.getId())
    .map(summary -> translateSummaryToTupleDomain(dynamicFilter.getId(), summary, sourceColumnHandles)))
  .filter(Optional::isPresent)
  .reduce(TupleDomain.all(), TupleDomain::intersect)
```